src/common/Utilities.py

```diff
@@ -464,28 +464,41 @@ def str_to_int_list(str_list):
         int_list = int_list + int_item
     return int_list
 
-def dot_product_test(operator, num_tests = 20, max_err = 10e-5):
+def test_operator_adjoint(operator, num_tests = 20, max_err = 10e-5):
     '''
     Test if a given operator is adjoint.
-    The operator needs to have been already set_up() with valid AcquisitionModel and ImageData objects.
+    The operator needs to have been already set_up() with valid objects.
     The operator needs to have methods direct() and adjoint() implemented
+
+    Parameters
+    ----------
+    operator  :
+        Any SIRF operator that implements direct() and adjoint()
+    num_tests : int, optional
+        Number of tests with random data that will be executed. Default 20
+    max_err   : double, optional
+        Maximum allowed normalized error, tolerance. Change not recommended. Default 10e-5
     '''
-    # Get image and data to know the sizes of each operation.
-    image=operator.domain_geometry()
-    image.fill(1.0)
-    data=operator.direct(image) # Is there a better way than this? Similar to .domain_geomerty() ?
+    # Get x to know the size of each operation (we can not currently get y, so it will be obtained in the loop)
+    x=operator.domain_geometry()
     for __ in range(num_tests):
-        # generate random data
-        x0 = numpy.random.rand(*image.shape) * 10**numpy.random.randint(-2, 3)
-        y0 = numpy.random.rand(*data.shape ) * 10**numpy.random.randint(-2, 3)
-        # fill image and data
-        image.fill(x0)
-        data.fill(y0)
-        # compute direct and adjoint
-        y_hat = operator.direct(image)
-        x_hat = operator.adjoint(data)
+        ## generate random data and direct()
+        x0 = numpy.random.rand(*x.shape) * 10**numpy.random.randint(-2, 3)
+        # add a complex part if necessary
+        if numpy.any(numpy.iscomplex(x.as_array())):
+            x0=x0+1j*numpy.random.rand(*x.shape) * 10**numpy.random.randint(-2, 3)
+        x.fill(x0)
+        y_hat = operator.direct(x)
+        ## generate random data and adjoint()
+        y0 = numpy.random.rand(*y_hat.shape) * 10**numpy.random.randint(-2, 3)
+        # add a complex part if necessary
+        if numpy.any(numpy.iscomplex(y_hat.as_array())):
+            y0=y0+1j*numpy.random.rand(*y_hat.shape) * 10**numpy.random.randint(-2, 3)
+        y=y_hat.copy()
+        y.fill(y0)
+        x_hat = operator.adjoint(y)
         # Check dot product identity
-        norm_err = abs(y_hat.dot(data) - x_hat.dot(image))/(y_hat.dot(data)*0.5 + x_hat.dot(image)*0.5)
+        norm_err = abs(y_hat.dot(y) - x_hat.dot(x))/(y_hat.dot(y)*0.5 + x_hat.dot(x)*0.5)
```

Comment:
would need complex conjugate when just using `dot` I guess. See e.g. https://mathworld.wolfram.com/HermitianInnerProduct.html

Also, in the denominator need `abs(y_hat.dot(data))*0.5 + ...`

Reply:
@KrisThielemans I believe this should be already handled by `AdquisitionData.dot()` and `ImageData.dot()`. I will make sure its like that.

Reply:
might be. If it does, it would probably make more sense to adhere to `numpy` conventions where `dot` is matrix multiplication, but `vdot` uses complex conjugate. https://docs.scipy.org/doc/numpy/reference/generated/numpy.vdot.html#numpy.vdot. Once you know what happens, please create an issue accordingly.

Reply:
@KrisThielemans hum, so according to those conventions, if this works, all of them should be `vdot` in SIRF, right?

Reply:
guess so. Please just raise a SIRF issue for the creation of `vdot`, but just use `dot` here (and `numpy` conjugate).

Reply:
@paskino this would also apply to CIL `dot_test`

Reply:
Sure. No need of numpy conjugate I believe. According to @paskino gadgetron.dot() will do it.

Reply:
then `gadgetron.dot` has the wrong name...

Reply:
@KrisThielemans  right, all of them should be `vdot`, as if its for numpy convention, all of them are wrong named, i.e. STIR does return matrices, not vectors, yet the `dot` does not compute matrix dot products, but vector dot products, right?

Reply:
but MATLAB uses `dot`, not `vdot`, and we are trying to make our Python and MATLAB scripts look as close as possible (ideally, ; at the end being the only difference).

Reply:
@evgueni-ovtchinnikov hum, but MATLAB's `dot` does not behave like neither  `np.dot` or `np.vdot`. The difference being that for matrices, MATLAB takes vector by vector dot product (vectorized approach to multiple `vdot`s) while python's `dot` does matrix multiplication.

Reply:
@evgueni-ovtchinnikov that's true, but in this case unfortunately we might have to diverge. We have to balance this with the "native" feeling of the wrapper. (another example being the index order).

In any case, MATLAB's `dot`  doesn't do complex conjugate either. https://uk.mathworks.com/help/matlab/ref/dot.html

Reply:
@KrisThielemans that is not correct. 

```
A = [1+i 1-i -1+i -1-i];
B = [3-4i 6-2i 1+2i 4+3i];

sum(conj(A).*B)
ans =1.0000 - 5.0000i

sum(A.*B)
ans = 7.0000 -17.0000i

dot(A,B)
ans =1.0000 - 5.0000i
```

Reply:
sorry for giving the wrong info. trying to be too fast!

so, we're stuck in naming now anyway. Seems that MATLAB's `dot` is neither like Python's `vdot` (it will give an n-1 dim array), nor `dot` (it does complex conjugate). sigh.

We might just have to give up and invent our own `scalar_product` function.

Let's see what @paskino thinks.

Reply:
My 2 cents: The behavior is closer to `vdot` than `dot` if we want to be closer to numpy. As this is python, I think it makes more sense to be closer to python implementations than MATLAB. In fact, what SIRF is doing in all cases is the same behavior as `vdot`: unroll, and complex conjugate if needed.

Reply:
I am marking this conversation as resolved, as it has drifted to `vdot` vs `dot`. Better to open a new issue on the topic.